cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/ObjectStorage.scala

```
@@ -85,4 +89,98 @@ object ObjectStorage {
     val directories: List[String]
     val listToken: Option[ListToken]
   }
+
+  final class ReaderInputStream private (reader: Reader) extends InputStream {
+    import ReaderInputStream._
+
+    private[this] var maybeBuffer = Option.empty[Try[Option[Array[Byte]]]]
```

The nested data structures of this value make some of the code harder to understand. I'm wondering why we need the `Try` and the outer `Option`?
- For the `Try`, can't we just let exceptions propagate when they are thrown? (all of the methods are invoked by the thread doing the reading, so it will be able to catch the exceptions as you wanted).
- For the outer `Option`, it looks like the only time when it's `None` is before any bytes have been read. Why not have the initial value be an empty array instead? That should have the same effect.

Summarizing:
```scala
private[this] var maybeBuffer: Option[Array[Byte]] = Some(Array.empty)
```